File: python/SortStacks.py
"""
Title: Sort stacks

Given a stack, sort the elements in the stack using one additional stack.

Execution: python SortStacks.py

For more details, check out http://www.byte-by-byte.com/sortstacks/
"""
import unittest


def sort_stacks(stack):
    if stack is None or stack == []:
        return stack
    new_stack = []
    new_stack.append(stack.pop())
    while stack != []:
        temp = stack.pop()
        while new_stack != [] and temp > new_stack[-1]:
            stack.append(new_stack.pop())
        new_stack.append(temp)
    return new_stack


class TestSortStacks(unittest.TestCase):
    def test_1(self):
        s = [1, 2, 3, 4, 5]
        s = sort_stacks(s)
        self.assertEqual(s, [5, 4, 3, 2, 1])

    def test_2(self):
        s = [5, 4, 3, 2, 1]
        s = sort_stacks(s)
        self.assertEqual(s, [5, 4, 3, 2, 1])

    def test_3(self):
        s = [-5, 4, 2, 1, 9]
        s = sort_stacks(s)
        self.assertEqual(s, [9, 4, 2, 1, -5])


if __name__ == '__main__':
    unittest.main()

File: java/SortStacks.java
/*
 *   Title: Sort stacks.
 *
 *   Given a stack, sort the elements in the stack using one additional stack.
 * 
 *   Execution: javac SortStacks.java && java SortStacks
 * 
 *   For more details, check out http://www.byte-by-byte.com/sortstacks
 */

import java.util.Stack;

public class SortStacks {
    public static Stack<Integer> sort(Stack<Integer> stack) {
        if (stack == null || stack.isEmpty()) {
            return stack;
        }
        Stack<Integer> newStack = new Stack<Integer>();
        newStack.push(stack.pop());
        while (!stack.isEmpty()) {
            int temp = stack.pop();
            while (!newStack.isEmpty() && temp > newStack.peek()) {
                stack.push(newStack.pop());
            }
            newStack.push(temp);
        }
        return newStack;
    }

    public static void test_1() {
        Stack<Integer> stack = new Stack<Integer>();
        stack.push(5);
        stack.push(3);
        stack.push(100);
        stack.push(-10);
        
        SortStacks ss = new SortStacks();
        Stack<Integer> sorted_stack = ss.sort(stack);

        for(Integer i : sorted_stack) {
            System.out.println(i);
        }
    }
    
    public static void test_2() {
        Stack<Integer> stack = new Stack<Integer>();
        stack.push(-10);
        stack.push(-5);
        stack.push(-20);
        stack.push(-50);
        
        SortStacks ss = new SortStacks();
        Stack<Integer> sorted_stack = ss.sort(stack);

        for(Integer i : sorted_stack) {
            System.out.println(i);
        }
    }

    public static void test_3() {
        Stack<Integer> stack = new Stack<Integer>();
        stack.push(5);
        stack.push(4);
        stack.push(3);
        stack.push(0);
        
        SortStacks ss = new SortStacks();
        Stack<Integer> sorted_stack = ss.sort(stack);

        for(Integer i : sorted_stack) {
            System.out.println(i);
        }
    }
    // Sample test cases
    public static void main(String[] args) {
        System.out.println("Test 1");
        test_1();

        System.out.println("Test 2");
        test_2();

        System.out.println("Test 3brew install vifm");
        test_3();
        System.out.println("Passed all test cases");
    }
}
